core/src/validator.rs

```diff
@@ -563,8 +568,71 @@ impl Validator {
     }
 }
 
+fn active_vote_account_exists_in_bank(bank: &Arc<Bank>, vote_account: &Pubkey) -> bool {
+    if let Some(account) = &bank.get_account(vote_account) {
+        if let Some(vote_state) = VoteState::from(&account) {
+            return !vote_state.votes.is_empty();
+        }
+    }
+    false
+}
+
+fn post_process_restored_tower(
+    restored_tower: crate::consensus::Result<Tower>,
+    validator_identity: &Pubkey,
+    vote_account: &Pubkey,
+    config: &ValidatorConfig,
+    ledger_path: &Path,
+    bank_forks: &BankForks,
+) -> Tower {
+    restored_tower
+        .and_then(|tower| {
+            let root_bank = bank_forks.root_bank();
+            let slot_history = root_bank.get_slot_history();
+            tower.adjust_lockouts_after_replay(root_bank.slot(), &slot_history)
+        })
+        .unwrap_or_else(|err| {
+            let is_already_active =
+                active_vote_account_exists_in_bank(&bank_forks.working_bank(), &vote_account);
```

Comment:
(repost) Hmm, maybe should this be `bank_forks.root_bank()`? Otherwise, can any bad validator trick victim validator with crafted block with high-number crafted slot, somehow? I mean the victim can be tricked into the wrong perception that the vote account doesn't exist.

Reply:
hmmmm so if the vote account exists in the `root bank()`, then unless the vote account was cleaned up, it should also exist in the `working_bank` because the `working_bank` must be a verified/replayed descendant of the `root_bank` right?

On a side note, it's a good question what we should do if validators' saved towers do get corrupted. Without slashing, should generally be ok for correctness (no rollback) if <33% experience a failure on restart at the same time

Reply:
> hmmmm so if the vote account exists in the `root bank()`, then unless the vote account was cleaned up .... verified/replayed ...

I now understand this isn't dangerous... I misunderstood that a victim validator could tricked that its vote account is missing by being fed with crafted blocks under fully network-controlled attack scenario. Well, even such an unlikely 
dire situation, a malicious validator can't remove victim's vote account without authorization unless the private key is leaked. (implied by _verified_ in the quote)

> On a side note, it's a good question what we should do if validators' saved towers do get corrupted. 

Yeah, that's interesting consideration... Maybe renew vote account and announce its stakers to migrate as well? Or should we introduce some quarantine period (1 epoch or so) to be free of slashing? Also, should I prepare doc update for general such emergency practice guideline? (I'll do as a separate pr...) Or punt it until slashing? ;)

> Without slashing, it may be ok to just start fresh again with a new tower ...

To check my understanding, are you meaning that rollback won't occur even if sizable portion of validators' saved tower get corrupted at the same time as long as the portion is less than superminority?

Reply:
> we introduce some quarantine period (1 epoch or so) to be free of slashing? Also, should I prepare doc update for general such emergency practice guideline? (I'll do

Yeah hehe I think that's practically what we'll have to do, whenever we do introduce slashing. Until then, we can't really make an informed decision :)

Reply:
> To check my understanding, are you meaning that rollback won't occur even if sizable portion of validators' saved tower get corrupted at the same time as long as the portion is less than superminority?

Yeah, I think as long as all failures  including dead validators, tower corruption < 33%, we shouldn't see rollback

Reply:
- [ ] todo: add logging by `error` measure?

Reply:
done! 207a9c9f03d31ed4b3dad91e3f45536a601b0339